includes/data-port/class-sensei-data-port-job

```diff
@@ -0,0 +1,347 @@
+	 */
+	abstract public static function create( $job_id, $sub_jobs );
+
+	/**
+	 * Do question porting.
+	 *
+	 * @param array $question_state  Question sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                               'completed' when the port is completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_question_port( &$question_state );
+
+	/**
+	 * Do lesson porting.
+	 *
+	 * @param array $lesson_state  Lesson sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                             'completed' when the port is completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_lesson_port( &$lesson_state );
+
+	/**
+	 * Do course porting.
+	 *
+	 * @param array $course_state  Course sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                             'completed' when the port id completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_course_port( &$course_state );
+
+	/**
+	 * Serialize state to JSON.
+	 *
+	 * @return array
+	 */
+	public function jsonSerialize() {
+		return [
+			's' => $this->sub_jobs_state,
+			'l' => $this->logs,
+			'r' => $this->results,
+			'c' => $this->is_completed,
+			'p' => $this->percentage,
+		];
+	}
+
+	/**
+	 * Restore state from JSON.
+	 */
+	private function restore_from_json( $json_string ) {
+		$json_arr = json_decode( $json_string, true );
+
+		if ( ! $json_arr ) {
+			return;
+		}
+
+		$this->sub_jobs_state = $json_arr['s'];
+		$this->logs           = $json_arr['l'];
+		$this->results        = $json_arr['r'];
+		$this->is_completed   = $json_arr['c'];
+		$this->percentage     = $json_arr['p'];
+	}
+
+	/**
+	 * Add an entry to the logs.
+	 *
+	 * @param string $name      Name of the entity this log applies to.
+	 * @param string $message   Log message.
+	 * @param string $id        Id of the entity this log applies to.
+	 */
+	protected function add_log_entry( $name, $message, $id = '' ) {
+		$this->has_changed = true;
+
+		$entry = [
+			'name' => sanitize_text_field( $name ),
+			'msg'  => sanitize_text_field( $message ),
+		];
+
+		if ( ! empty( $id ) ) {
+			$entry['id'] = sanitize_text_field( $id );
+		}
+
+		$this->logs[] = $entry;
+	}
+
+	/**
+	 * Persist state to the db.
+	 */
+	public function persist() {
+		if ( $this->has_changed ) {
+			update_option( self::get_option_name(), wp_slash( wp_json_encode( $this ) ) );
+		}
+	}
+
+	/**
+	 * Run the job.
+	 */
+	public function run() {
+		if ( $this->is_completed || empty( $this->sub_jobs_state ) ) {
+			return;
+		}
+
+		$this->has_changed = true;
+		$this->percentage  = 0;
+		$subjob_percentage = 1 / count( $this->sub_jobs_state );
+
+		if ( ! empty($this->sub_jobs_state[ 'questions' ][ 'status' ]) ) {
+
+			if ( $this->sub_jobs_state[ 'questions' ][ 'status' ] !== 'completed' ) {
```

Comment:
I think we can define these in the child classes and loop through sub-jobs/tasks. They will be slightly different for import/export. I wonder if we should also have a way to resume mid-way a task, but maybe that will be in the sub-job/task state. We’ll want to have the ability to use pagination on this job.

Reply:
Yes the idea is that since each task has its own state, it can store whatever it wants in that state in order to resume and implement pagination. In my first approach I thought implementing something in the superclass but I decided against it. With this approach the subclass' only responsibility is to mark itself as completed and any differences can be handled in the subclass itself.